feat: Add EventStream service abstraction

mattrobenolt opened the pull request:
This is bringing in the code we had specific to getsentry into a configurable backend for our EventStream abstraction. This also will open up the doors for the kafka based consumer coming soon and a development/local version that sends data directly to Snuba bypassing kafka.

This needs to be coordinated with a PR on getsentry side before merging, otherwise we will break stuff.

mattrobenolt committed:
feat: Add EventStream service abstraction

mattrobenolt committed:
Base publish just directly triggers post_processing_group

mattrobenolt committed:
Add consume abstraction temporarily

mattrobenolt committed:
s/topic/publish_topic/

mattrobenolt commented on src/sentry/eventstream/kafka.py:
```diff
@@ -0,0 +1,70 @@
+from __future__ import absolute_import
```
This is pretty much yanked directly from https://github.com/getsentry/getsentry/blob/master/getsentry/processing.py with a few modifications to be a service.

tkaemming replied:
👍

mattrobenolt commented on src/sentry/eventstream/base.py:
```diff
@@ -0,0 +1,17 @@
+        'publish', 'consume',
+    )
+
+    def publish(self, event, primary_hash, **kwargs):
```
These signatures are effectively temporary until @tkaemming rolls in and changes things up.

tkaemming replied:
So, this actually won't work with the delegator right now due to the `**kwargs`: https://github.com/getsentry/sentry/blob/53837d9e8dad2443c3881fcd05f1b261bdd922fd/src/sentry/utils/services.py#L285-L293

It's because this doesn't hold:

```python
>>> import inspect
>>> f = lambda x, y, *a, **k: (x, y, a, k)
>>> f(0, 1, 2, a=3, b=4) == inspect.getcallargs(f, 0, 1, 2, a=3, b=4)
False
```

I'm going to need to fix that regardless but if you want to get this out before I have a solution, it'd be better to just make all of these arguments explicit in the function definition. I think what you have now is fine in `src/sentry/event_manager.py`.

mattrobenolt committed:
lol

mattrobenolt committed:
Fix tests

tkaemming commented on src/sentry/eventstream/base.py:
```diff
@@ -0,0 +1,17 @@
+
+class EventStream(Service):
+    __all__ = (
+        'publish', 'consume',
```
I think you were right yesterday when you mentioned only providing one of the methods to start and then filling in the other one later. I'm not exactly sure how the Kafka consumer end of this is going to look yet, so I think it would probably make sense to only define the `publish` method for now?

mattrobenolt committed:
Remove kwargs and consume method

tkaemming reviewed (approved):
🍪

tkaemming commented:
If you need help getting the delegator stuff going to run both in production, let me know — I think it should be pretty straightforward.

mattrobenolt committed:
Push raw events also through eventstream, but flag them to not be consumed

This "raw" feature is only used by `create_sample_event`, and is meant
to be used to bypass notifications and other post processing events.

Prior to this patch, this meant that all sample events were skipping the
eventstream entirely. This is a concern for both production (since you
can trigger sample events through the UI) and local for load-mocks,
which bypass being stored into the eventstream.

mattrobenolt commented:
@tkaemming Wanna take a look at https://github.com/getsentry/sentry/pull/8879/commits/312460cc90dd670c39a17d85204d07764288bb72 ?

Sorry, was in the middle of finishing that up while you approved it.

tkaemming reviewed (approved):
Makes sense. I don't know that `consume` is the clearest word to use here but I don't have any better ideas.
